lib/std/crypto/rsa.zig

```diff
@@ -0,0 +1,731 @@
+
+                if (msg.len > k - 2 * hLen - 2) return error.MessageTooLong;
+
+                var db = [_]u8{0} ** (k - hLen - 1);
+                @memcpy(db[0..lHash.len], &lHash);
+                db[db.len - msg.len - 1] = 1;
+                @memcpy(db[db.len - msg.len ..], msg);
+
+                var seed: [Hash.digest_length]u8 = undefined;
+                std.crypto.random.bytes(&seed);
+
+                const db_mask = mgf1(Hash, &seed, db.len);
+                for (&db, db_mask) |*v, m| v.* ^= m;
+
+                const seed_mask = mgf1(Hash, &db, hLen);
+                for (&seed, seed_mask) |*v, m| v.* ^= m;
+
+                var em = [_]u8{0} ++ seed ++ db;
+
+                const m = Fe.fromBytes(pk.modulus, &em, .big) catch unreachable;
+                const e = pk.modulus.powPublic(m, pk.public_exponent) catch unreachable;
+                e.toBytes(&em, .big) catch unreachable;
+                return em;
+            }
+        };
+
+        pub const SecretKey = struct {
+            /// `d`
+            private_exponent: Fe,
+
+            pub fn fromBytes(n: Modulus, exp: []const u8) !SecretKey {
+                const d = try Fe.fromBytes(n, exp, .big);
+                if (std.debug.runtime_safety) {
+                    // > The RSA private exponent d is a positive integer less than n
+                    // > satisfying e * d == 1 (mod \lambda(n)),
+                    if (!d.isOdd()) return error.InvalidExponent;
+                    if (d.v.compare(n.v) != .lt) return error.InvalidExponent;
+                }
+
+                return .{ .private_exponent = d };
+            }
+        };
+
+        pub const KeyPair = struct {
+            public: PublicKey,
+            secret: SecretKey,
+
+            pub fn fromDer(parser: *der.Parser) !KeyPair {
+                // We're just interested in the first few fields which don't vary by version
+                _ = try parser.expectSequence();
+                const version = try parser.expectPrimitive(.integer);
+                _ = version;
+
+                const mod = try parseModulus(parser);
+                const pub_exp = try parseInteger(parser);
+                const sec_exp = try parseInteger(parser);
+                // Skip unused private key fields
+
+                const public = try PublicKey.fromBytes(mod, pub_exp);
+                const secret = try SecretKey.fromBytes(public.modulus, sec_exp);
+
+                // TODO: make this work (FeUnit multiplication?)
+                if (std.debug.runtime_safety) {
+                    // const prime1 = try parseInteger(&parser);
+                    // const prime2 = try parseInteger(&parser);
+                    // const p = try Fe.fromBytes(public.modulus, prime1, .big);
+                    // const q = try Fe.fromBytes(public.modulus, prime2, .big);
+
+                    // // check that n = p * q
+                    // const n_expected = p.v.mul(q.v); // can't mul these :(
```

Comment:
Oh, could use your help with this @jedisct1 . Is there a way to multiple these, but not under the modulus?

Reply:
Why not under the modulus? `p*q (mod n)` should be zero.

Reply:
BTW, don't assume that there are only two primes.

Reply:
I found this check in OpenSSL, which I believe only checks the first two primes.